Only in rd/etc: rc
diff -ru rd-orig/etc/rc.network rd/etc/rc.network
--- rd-orig/etc/rc.network	2014-12-14 21:28:03.771087000 +0000
+++ rd/etc/rc.network	2014-12-15 22:40:54.491087000 +0000
@@ -76,7 +76,7 @@
 {
 	local LEASE_FILE_NAME="dhcpd.conf.leases"
 	local LEASE_FILE="dhcp-leasefile=/etc/dhcpd/${LEASE_FILE_NAME}"
-	local CONF_FILE=`ls /etc/dhcpd/dhcpd-*.conf`
+	local CONF_FILE=`ls /etc/dhcpd/dhcpd-*.conf 2> /dev/null`
 
 	if [ "${CONF_FILE}" = "" ]; then
 		return 1
@@ -90,7 +90,7 @@
 
 merge_dhcpd_info ()
 {
-	local CONF_FILE=`ls /etc/dhcpd/dhcpd-*.conf`
+	local CONF_FILE=`ls /etc/dhcpd/dhcpd-*.conf 2> /dev/null`
 	if [ "${CONF_FILE}" = "" ]; then
 		echo "enable=\"no\"" > ${DHCPD_INFO}
 		return 1
@@ -115,8 +115,8 @@
 	# the soft link will be generated, and the link should be remove when
 	# DHCP Stop.
 	if [ -e ${DHCP_PKG_ENABLE} ]; then
-		rm ${DHCPD_LEASES}
-		rm ${DHCPD_LEASES_LOG}
+		rm -f ${DHCPD_LEASES}
+		rm -f ${DHCPD_LEASES_LOG}
 	fi
 	return 0
 }
@@ -313,6 +313,7 @@
 		${S01Scripts} unload_nat_mod
 	fi
 
+	[ -e /etc/dhcpd/dhcpd.info ] && \
 	grep "enable=\"yes\"" /etc/dhcpd/dhcpd.info
 	if [ ! $? -eq 0 ]; then
 		killall dnsmasq
@@ -991,11 +992,11 @@
 
 		# setup mac addresses earlier than other model because
 		# dualhead model first interface is HA's heartbeat(hidden from UI)
-		if ! dualhead_is_heartbeat_ifn_exist; then
-			for ifn in ${network_interfaces}; do
-				set_mac_address ${ifn}
-			done
-		fi
+		#if ! dualhead_is_heartbeat_ifn_exist; then
+		#	for ifn in ${network_interfaces}; do
+		#		set_mac_address ${ifn}
+		#	done
+		#fi
 		dualhead_start_network "${network_interfaces}"
 	fi
 
@@ -1007,9 +1008,9 @@
 			continue
 		fi
 
-		if [ "yes" != "${SUPPORT_DUAL_HEAD}" ]; then
-			set_mac_address ${ifn}
-		fi
+		#if [ "yes" != "${SUPPORT_DUAL_HEAD}" ]; then
+		#	set_mac_address ${ifn}
+		#fi
 		eval bond_opt=\$Bond_${ifn}
 		[ -n "${bond_opt}" ] && install_bonding_driver ${ifn} "${bond_opt}"
 
diff -ru rd-orig/etc/rc.subr rd/etc/rc.subr
--- rd-orig/etc/rc.subr	2014-12-14 21:28:04.015087000 +0000
+++ rd/etc/rc.subr	2014-12-14 22:29:29.351087000 +0000
@@ -56,7 +56,7 @@
 	fi
 }
 
-: ${COLORIZED:=}
+: ${COLORIZED:=yes}
 _MsgInit
 
 MsgWarn()
diff -ru rd-orig/etc/rc.volume rd/etc/rc.volume
--- rd-orig/etc/rc.volume	2014-12-14 21:28:04.275087000 +0000
+++ rd/etc/rc.volume	2014-12-14 22:30:02.039087000 +0000
@@ -156,6 +156,7 @@
 
 	#Fix super block archive version issue.
 	#Do not remove it, the program will check archive version when reboot.
+	[ -e /usr/syno/bin/synoarchiveupgrade ] && \
 	/usr/syno/bin/synoarchiveupgrade &
 
 	#Recommend users to do data scrubbing if needed
diff -ru rd-orig/linuxrc.syno rd/linuxrc.syno
--- rd-orig/linuxrc.syno	2014-12-14 21:28:04.519087000 +0000
+++ rd/linuxrc.syno	2014-12-14 22:42:29.035087000 +0000
@@ -1,7 +1,7 @@
 #!/bin/sh
 # Copyright (c) 2000-2014 Synology Inc. All rights reserved.
 
-echo "START /linuxrc.syno"
+RCMsg "START /linuxrc.syno" true
 
 . /etc.defaults/rc.subr
 
@@ -26,6 +26,10 @@
 USB_MODULES="${USB_MODULES} etxhci-hcd" # for Etron USB3.0
 SupportDualhead=`/bin/get_key_value $SYNOINFO_DEF support_dual_head`
 
+#### ---- XPEnology ---- ####
+[ -f /.gnoboot/rc.d/pre ] && source /.gnoboot/rc.d/pre
+#### ---- XPEnology ---- ####
+
 Exit()
 {
 	if [ -n "$2" ]; then
@@ -35,9 +39,9 @@
 	# show date for login info
 	date
 
-	grep "^${RootDevice}" /proc/mounts && /bin/umount -f ${Mnt}
-	grep "sysfs" /proc/mounts && /bin/umount /sys
-	grep "^/proc" /proc/mounts && /bin/umount /proc
+	grep -q "^${RootDevice}" /proc/mounts && /bin/umount -f ${Mnt}
+	grep -q "sysfs" /proc/mounts && /bin/umount /sys
+	grep -q "^/proc" /proc/mounts && /bin/umount /proc
 	exit $1
 }
 
@@ -48,7 +52,7 @@
 if [ -x /usr/syno/bin/synocfgen ]; then
         insmod /lib/modules/synobios.*
         /bin/mknod /dev/synobios c 201 0
-        echo "Starting /usr/syno/bin/synocfgen..."
+        RCMsg "Starting /usr/syno/bin/synocfgen..."
         /usr/syno/bin/synocfgen
         RetCfg=$?
         echo "/usr/syno/bin/synocfgen returns $RetCfg"
@@ -239,7 +243,7 @@
 		echo "${FSCK} returns ${ResFSCK}"
 		case "$ResFSCK" in
 		4 | 8)
-			echo "${FSCK} returns ${ResFSCK}, move into network install..."
+			RCMsg "${FSCK} returns ${ResFSCK}, move into network install..."
             FsckFailed=1;
 		;;
 		esac
@@ -293,6 +297,7 @@
 #
 if [ 0 -eq $FsckFailed ]; then
     echo '------------upgrade'
+    RCMsg "Checking upgrade file" \
     /bin/sh /etc/upgrade.sh
     echo '============upgrade'
 fi
@@ -318,17 +323,17 @@
 fi
 
 if [ ! -d ${Mnt}/etc ]; then
-	echo 'Copy default /etc'
+	RCMsg 'Copy default /etc'
 	rm -rf ${Mnt}/etc
 	cp -R ${Mnt}/etc.defaults ${Mnt}/etc
 fi
 if [ ! -d ${Mnt}/usr/syno/etc ]; then
-	echo 'Copy default /usr/syno/etc'
+	RCMsg 'Copy default /usr/syno/etc'
 	rm -rf ${Mnt}/usr/syno/etc
 	cp -R ${Mnt}/usr/syno/etc.defaults ${Mnt}/usr/syno/etc
 fi
 if [ ! -d ${Mnt}/var ]; then
-	echo 'Copy default /var'
+	RCMsg 'Copy default /var'
 	rm -rf ${Mnt}/var
 	cp -R ${Mnt}/var.defaults ${Mnt}/var
 fi
@@ -353,6 +358,10 @@
 	Exit 4 ".memtest exists"
 fi
 
+#### ---- XPEnology ---- ####
+[ -f /.gnoboot/rc.d/post ] && source /.gnoboot/rc.d/post
+#### ---- XPEnology ---- ####
+
 UniqueFlash=`/bin/get_key_value /etc/synoinfo.conf unique`
 UniqueHD=`/bin/get_key_value ${Mnt}/etc.defaults/synoinfo.conf unique`
 
@@ -360,12 +369,12 @@
 	Exit 4 "Unique not match"
 fi
 
-VersionFlash=`/bin/get_key_value /etc.defaults/VERSION buildnumber`
-VersionHD=`/bin/get_key_value ${Mnt}/etc.defaults/VERSION buildnumber`
+#VersionFlash=`/bin/get_key_value /etc.defaults/VERSION buildnumber`
+#VersionHD=`/bin/get_key_value ${Mnt}/etc.defaults/VERSION buildnumber`
 
-if [ "$VersionFlash" != "$VersionHD" ]; then
-	Exit 7 "Version number not match"
-fi
+#if [ "$VersionFlash" != "$VersionHD" ]; then
+#	Exit 7 "Version number not match"
+#fi
 
 if [ 1 -eq $FsckFailed ]; then
 	Exit 5 "${RootDevice} fsck fail or had lost\+found files"
